imap_processing/mag/l1a/mag_l1a_data.py

```diff
@@ -429,6 +477,463 @@ def to_signed16(n: int) -> int:
 
         return (
-            np.array(primary_vectors, dtype=np.int64),
-            np.array(secondary_vectors, dtype=np.int64),
+            np.array(primary_vectors, dtype=np.int32),
+            np.array(secondary_vectors, dtype=np.int32),
+        )
+
+    @staticmethod
+    def process_compressed_vectors(  # noqa: PLR0912, PLR0915
+        vector_data: np.ndarray, primary_count: int, secondary_count: int
+    ) -> tuple[np.ndarray, np.ndarray]:
+        """
+        Given raw compressed packet data, process into Vectors.
+
+        To do this, we need to decode the compressed data. The compressed data starts
+        with an 8 bit header that defines the width of the uncompressed vectors and
+        if there is a range data section. Then, the vector data follows, then the range
+        data section if it exists.
+
+        To decode, we start by decoding the first compression_width bits. This is an
+        uncompressed primary vector with range. Then, we proceed through the compressed
+        data, where each value is fibonacci and zig-zag encoded. This means each value
+        ends in 2 sequential ones (11). We split the data along these numbers until
+        we reach primary_count.
+
+        The secondary vectors are decoded the same way, starting directly after the last
+        primary vector with an uncompressed secondary starting vector and then
+        secondary_count compressed vectors.
+
+        The compressed values are differences from the previous vector, so after
+        decoding we accumulate the values starting from the first known vector. The
+        range data is copied from the starting vector if range_data_section is not
+        included.
+
+        Then, if a range data section is included, we decode it and assign it to each
+        vector. There are 2 * (primary_count + secondary_count) bits assigned for the
+        range data section.
+
+        If any compressed vectors are > 60 bits long, then we switch to uncompressed
+        vectors for the rest of the processing.
+
+        Parameters
+        ----------
+        vector_data : numpy.ndarray
+            Raw vector data, in bytes. Contains both primary and secondary vector data.
+        primary_count : int
+            Count of the number of primary vectors.
+        secondary_count : int
+            Count of the number of secondary vectors.
+
+        Returns
+        -------
+        (primary, secondary): (numpy.ndarray, numpy.ndarray)
+            Two arrays, each containing tuples of (x, y, z, sample_range) for each
+            vector sample.
+        """
+        bit_array = np.unpackbits(vector_data)
+        # The first 8 bits are a header - 6 bits to indicate the compression width,
+        # 1 bit to indicate if there is a range data section, and 1 bit spare.
+        compression_width = int("".join([str(i) for i in bit_array[:6]]), 2)
+        has_range_data_section = int(str(bit_array[6]), 2)
```

Comment:
```suggestion
        has_range_data_section = (bit_array[6] == 1)
```
be nicer if this was typed as a bool?